File: source/1049.最后一块石头的重量-ii.cpp
/*
 * @lc app=leetcode.cn id=1049 lang=cpp
 *
 * [1049] 最后一块石头的重量 II
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
public:
    int lastStoneWeightII(vector<int>& stones) {
        vector<int> f(1001, 0);
        int sum = accumulate(stones.begin(), stones.end(), 0);
        int target = sum / 2;
        for (int i = 0; i < stones.size(); i++) {
            for (int j = target; j >= stones[i]; j--) {
                f[j] = max(f[j], f[j-stones[i]]+stones[i]);
            }
        }
        return sum - f[target]-f[target];
    }
};
// @lc code=end

File: 1049.最后一块石头的重量-ii.py
#
# @lc app=leetcode.cn id=1049 lang=python
#
# [1049] 最后一块石头的重量 II
#

# @lc code=start
class Solution(object):
    def lastStoneWeightII(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        target = sum(stones)/2.0
        candidates = {0}

        for x in stones:
            addition = set()
            for y in candidates:
                s = x + y
                if s == target:
                    return 0
                elif x+y < target:
                    addition.add(s)
            candidates = candidates.union(addition)
        return int(2*(target-max(candidates)))
# @lc code=end

